File: BHAVRP_Python/cujae/inf/citi/om/heuristic/assignment/classical/urgency/parallel_plus.py
import numpy as np
from typing import List
from parallel import Parallel
from .....problem.input.problem import Problem
from .....problem.input.customer import Customer
from .....problem.output.solution.solution import Solution
from .....problem.output.solution.cluster import Cluster

class ParallelPlus(Parallel):
    
    def __init__(self):
        super().__init__()
        
    def to_clustering(self) -> Solution:
        solution = Solution()
        
        list_clusters: List[Cluster] = self.initialize_clusters()
        list_customers_to_assign: List[Customer] = list(Problem.get_problem().get_customers().copy())
        list_id_depots: List[int] = list(Problem.get_problem().get_list_id_depots())
        
        urgency_matrix: np.ndarray = np.array(Problem.get_problem().get_cost_matrix())
        closest_matrix: np.ndarray = np.array(Problem.get_problem().get_cost_matrix())
        
        list_depots_ordered: List[List[int]] = self.get_depots_ordered(
            list_customers_to_assign, list_id_depots, closest_matrix
        )
        list_urgencies: List[float] = self.get_list_urgencies(
            list_customers_to_assign, list_depots_ordered, urgency_matrix, -1
        )
        
        while list_customers_to_assign and list_clusters:
            pos_customer = self.get_pos_max_value(list_urgencies)
            customer = list_customers_to_assign[pos_customer]
            request_customer = customer.get_request_customer()

            id_closest_depot = list_depots_ordered[pos_customer][0]
            capacity_depot = Problem.get_problem().get_total_capacity_by_depot(
                Problem.get_problem().get_depot_by_id_depot(id_closest_depot)
            )

            pos_cluster = self.find_cluster(id_closest_depot, list_clusters)

            if pos_cluster != -1:
                request_cluster = list_clusters[pos_cluster].get_request_cluster()

                if capacity_depot >= (request_cluster + request_customer):
                    request_cluster += request_customer

                    list_clusters[pos_cluster].set_request_cluster(request_cluster)
                    list_clusters[pos_cluster].get_items_of_cluster().append(customer.get_id_customer())

                    list_customers_to_assign.pop(pos_customer)
                    list_urgencies.pop(pos_customer)
                    list_depots_ordered.pop(pos_customer)

                customers_out_depot = self.get_customers_out_depot(
                    list_customers_to_assign, request_cluster, capacity_depot
                )

                if customers_out_depot:
                    if len(customers_out_depot) == len(list_customers_to_assign):
                        list_id_depots.remove(
                            Problem.get_problem().find_pos_element(list_id_depots, id_closest_depot))

                        if list_clusters[pos_cluster].get_items_of_cluster():
                            solution.get_clusters().append(list_clusters.pop(pos_cluster))
                        else:
                            list_clusters.pop(pos_cluster)
                            
                    for i in customers_out_depot:
                        current_pos_depot = Problem.get_problem().find_pos_element(
                            list_depots_ordered[i], id_closest_depot)

                        if current_pos_depot != -1:
                            list_depots_ordered[i].pop(current_pos_depot)

                            if not list_depots_ordered[i]:
                                solution.get_unassigned_items().append(list_customers_to_assign[i].get_id_customer())

                                list_customers_to_assign.pop(i)
                                list_urgencies.pop(i)
                                list_depots_ordered.pop(i)
                            else:
                                list_urgencies[i] = self.get_surgency(
                                    list_customers_to_assign[i].get_id_customer(),
                                    list_depots_ordered[i],
                                    urgency_matrix, -1)

        if list_customers_to_assign:
            for customer in list_customers_to_assign:
                solution.get_unassigned_items().append(customer.get_id_customer())

        if list_clusters:
            for cluster in list_clusters:
                if cluster.get_items_of_cluster():
                    solution.get_clusters().append(cluster)

        return solution

File: BHAVRP_Java/BHAVRP2.0/src/cujae/inf/ic/om/heuristic/assignment/classical/urgency/ParallelPlus.java
package cujae.inf.ic.om.heuristic.assignment.classical.urgency;

import java.util.ArrayList;

import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Problem;

import cujae.inf.ic.om.problem.output.solution.Cluster;
import cujae.inf.ic.om.problem.output.solution.Solution;

import cujae.inf.ic.om.matrix.NumericMatrix;

public class ParallelPlus extends Parallel {

	public ParallelPlus() {
		super();
	}
	public Solution toClustering(){
		Solution solution = new Solution();		
		
		ArrayList<Cluster> clusters = initializeClusters();	

		ArrayList<Customer> customersToAssign = new ArrayList<Customer>(Problem.getProblem().getCustomers());
		ArrayList<Integer> listIDDepots = new ArrayList<>(Problem.getProblem().getListIDDepots());
		
		NumericMatrix urgencyMatrix = new NumericMatrix(Problem.getProblem().getCostMatrix());
		NumericMatrix closestMatrix = new NumericMatrix(Problem.getProblem().getCostMatrix());

		ArrayList<ArrayList<Integer>> listDepotsOrdered = getDepotsOrdered(customersToAssign, listIDDepots, closestMatrix);
		ArrayList<Double> listUrgencies = getListUrgencies(customersToAssign, listDepotsOrdered, urgencyMatrix, -1);

		Customer customer = new Customer();
		double requestCustomer = 0.0;
		int posCustomer = -1;
		
		int idClosestDepot = -1;
		double capacityDepot = 0.0;
		
		int posCluster = -1;
		double requestCluster = 0.0;

		while((!customersToAssign.isEmpty()) && (!clusters.isEmpty())) 
		{
			posCustomer = getPosMaxValue(listUrgencies);
			customer = customersToAssign.get(posCustomer);
			requestCustomer = customer.getRequestCustomer();

			idClosestDepot = listDepotsOrdered.get(posCustomer).get(0).intValue(); 
			capacityDepot = Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepotByIDDepot(idClosestDepot));

			posCluster = findCluster(idClosestDepot, clusters);

			if(posCluster != -1)
			{
				requestCluster = clusters.get(posCluster).getRequestCluster();

				if(capacityDepot >= (requestCluster + requestCustomer))
				{	
					requestCluster += requestCustomer;

					clusters.get(posCluster).setRequestCluster(requestCluster);
					clusters.get(posCluster).getItemsOfCluster().add(customer.getIDCustomer());

					customersToAssign.remove(posCustomer);
					listUrgencies.remove(posCustomer);
					listDepotsOrdered.remove(posCustomer);
				}	

				ArrayList<Integer> customersOutDepot = getCustomersOutDepot(customersToAssign, requestCluster, capacityDepot);
				int currentPosDepot = -1;

				if(!customersOutDepot.isEmpty())
				{
					if(customersOutDepot.size() ==  customersToAssign.size())
					{
						listIDDepots.remove(Problem.getProblem().findPosElement(listIDDepots, idClosestDepot));

						if(!(clusters.get(posCluster).getItemsOfCluster().isEmpty()))
							solution.getClusters().add(clusters.remove(posCluster));
						else
							clusters.remove(posCluster);
					}

					for(int i = 0; i < customersOutDepot.size(); i++)
					{ 
						currentPosDepot = Problem.getProblem().findPosElement(listDepotsOrdered.get(customersOutDepot.get(i)), idClosestDepot);

						if(currentPosDepot != -1)
						{
							listDepotsOrdered.get(customersOutDepot.get(i)).remove(currentPosDepot);

							if(listDepotsOrdered.get(customersOutDepot.get(i)).isEmpty())
							{
								solution.getUnassignedItems().add(customersToAssign.get(customersOutDepot.get(i)).getIDCustomer());

								customersToAssign.remove(customersOutDepot.get(i));
								listUrgencies.remove(customersOutDepot.get(i));
								listDepotsOrdered.remove(customersOutDepot.get(i));
							}
							else
								listUrgencies.set(customersOutDepot.get(i), getUrgency(customersToAssign.get(customersOutDepot.get(i)).getIDCustomer(), listDepotsOrdered.get(customersOutDepot.get(i)), urgencyMatrix, -1));
						}	
					}
				}
			}
		}
		
		if(!customersToAssign.isEmpty())					
			for(int j = 0; j < customersToAssign.size(); j++)	
				solution.getUnassignedItems().add(customersToAssign.get(j).getIDCustomer());

		if(!clusters.isEmpty())
			for(int k = 0; k < clusters.size(); k++)
				if(!(clusters.get(k).getItemsOfCluster().isEmpty()))
					solution.getClusters().add(clusters.get(k));

		return solution;
	}
}